hip_data_tools/aws/s3.py

```diff
@@ -190,3 +181,239 @@ def _list_object_page(self, key_prefix, continuation_token):
             Prefix=key_prefix,
             ContinuationToken=continuation_token,
         )
+
+    def upload_directory(self,
+                         source_directory: str,
+                         extension: str,
+                         target_key: str,
+                         overwrite: bool = True,
+                         rename: bool = True) -> None:
+        """
+        Upload a local file directory to s3
+        Args:
+            source_directory (str): Local source directory's absolute path.
+            extension (str): the file extension of files in that directory to be uploaded.
+            target_key (str): Target location on the s3 bucket for files to be uploaded.
+            overwrite (bool): overwrite files on s3 or not
+            rename (bool): rename the file when uploading to s3 or not
+        Returns: None
+        """
+        if overwrite:
+            log.info("Cleaning existing flies on s3")
+            self.delete_recursive(f"{target_key}/")
+        log.info("searching for files to upload in %s", source_directory)
+        path_list = Path(source_directory).glob(f'**/*.{extension}')
+        itr = 0
+        upload_data = []
+        for path in path_list:
+            path_in_str = str(path)
+            filename = os.path.basename(path_in_str)
+            if rename:
+                filename = f"file-{str(uuid.uuid4())}.{extension}"
+            destination_key = f"{target_key}/{filename}"
+            itr = itr + 1
+            upload_data += [(self.conn.settings, path_in_str, self.bucket, destination_key)]
+        pool_size = min(16, max(1, int(len(upload_data) / 3)))  # limit pool size between 1 and 16
```

Comment:
good question, this was initially aimed at a certain band of ec2 machines, but since this is now a general purpose library, I'll make it generic Pool to consume all cores.